packages/backend/src/outgoing_payment/lifecycle.ts

```diff
@@ -0,0 +1,341 @@
+import * as assert from 'assert'
+import * as Pay from '@interledger/pay'
+import { debounce } from 'debounce'
+import { OutgoingPayment, PaymentState } from './model'
+import { ServiceDependencies } from './service'
+import { IlpPlugin } from './ilp_plugin'
+
+// Minimum interval between progress updates.
+const PROGRESS_UPDATE_INTERVAL = 200 // milliseconds
+const MAX_INT64 = BigInt('9223372036854775807')
+
+export type PaymentError = LifecycleError | Pay.PaymentError
+export enum LifecycleError {
+  QuoteExpired = 'QuoteExpired',
+  // Rate fetch failed.
+  PricesUnavailable = 'PricesUnavailable',
+  // Payment aborted via "cancel payment" API call.
+  CancelledByAPI = 'CancelledByAPI',
+  // Not enough money in the super-account.
+  InsufficientBalance = 'InsufficientBalance',
+  // Error from the account service, except an InsufficientBalance. (see: CreditError)
+  AccountServiceError = 'AccountServiceError',
+
+  // These errors shouldn't ever trigger (impossible states), but they exist to satisfy types:
+  MissingQuote = 'MissingQuote',
+  InvalidRatio = 'InvalidRatio'
+}
+
+export async function handleQuoting(
```

Comment:
> what is the expected behavior if two quotes come in for the same Outgoing Payment?

The `payment` passed to `handleQuoting` is already locked (see `worker.ts`, `getPendingPayment`) so other instances can't load & work on it simultaneously. Or do you mean if two different payments are created for the same invoice?

> Should you be able to call handleQuoting for an outgoing payment that has already had a quote?

If a payment has already been quoted, its state is changed, and `handlePaymentLifecycle` will not dispatch it to `handleQuoting` (except in the case of retries, where the new quote replaces the old one).

> Also I assume we should at a minimum do a check the payment state to not be sending or cancelled to ensure the quote doesn't inadvertently update an outgoing payment when its already been approved or being sent

`getPendingPayment` `SELECT`s a payment which is ready to be worked on and dispatches it to the correct handler in `lifecycle.ts`. For the duration of work the payment is locked (via transaction + `FOR UPDATE`).

Reply:
@sentientwaffle that makes sense.

It may be worth adding comments to the top of all these functions saying that the `knex` object being passed in is actually a transaction with a lock already taken out on the payment